### PR TITLE
Modify model appending algorithm

### PR DESCRIPTION
**Pull request by florent-lamiraux**
This modification follows https://github.com/stack-of-tasks/pinocchio/discussions/2203.
  
In the new version, when appending model B to model A on frame F, joints are copied in the
following order
  - joints of model A up to the parent of F,
  - all the descendents of parent of F,
  - the remaining joints of A.

### CODE REVIEW

File: include/pinocchio/algorithm/model.hxx [hunk 1]
```diff
@@ -221,6 +221,24 @@ namespace pinocchio
...
+    if (child == parent) return true;
+    JointIndex i = child;
+    while (i != 0)
+    {
+      i = model.parents[i];
+      if (i == parent) return true;
+    }
+    return false;
+  }
+
```
Comment (jcarpent) [inline include/pinocchio/algorithm/model.hxx:241]: Could you rather use `model.supports[child]` to check whether parent is supporting joint of `child`?

File: include/pinocchio/algorithm/model.hxx [hunk 2]
```diff
@@ -256,9 +274,28 @@ namespace pinocchio
...
+      if (hasAncestor(modelA, jid, frame.parent))
+      {
+        AJointsBeforeB.push_back(jid);
+      } else
+      {
+        AJointsAfterB.push_back(jid);
+      }
+    }
+    // Copy modelA joints that should come before model B
+    for (auto jid : AJointsBeforeB)
```
Comment (jcarpent) [inline include/pinocchio/algorithm/model.hxx:298]: This is a C++11 feature, targeted for Pinocchio 3.

File: include/pinocchio/algorithm/model.hxx [hunk 3]
```diff
@@ -275,7 +312,8 @@ namespace pinocchio
     }
 
     // Copy remaining joints of modelA
-    for (JointIndex jid = frame.parent+1; jid < modelA.joints.size(); ++jid)
+    // Copy modelA joints that should come before model B
+    for (auto jid : AJointsAfterB)
```
Comment (jcarpent) [inline include/pinocchio/algorithm/model.hxx:316]: Same comment apply here.

**Review by jcarpent — CHANGES_REQUESTED** [review]
Thanks a lot @florent-lamiraux for working on this feature.
I made some remarks to enhance the integration of this PR into Pinocchio.

**Comment by florent-lamiraux** [comment]
@jcarpent : I will handle your comments after fixing the tests.

File: include/pinocchio/algorithm/model.hxx (hunk 1 above)
Comment (florent-lamiraux) [reply include/pinocchio/algorithm/model.hxx:241]: It seems to me that the documentation of `model.supports` is incorrect. I would expect:
"supports[j] corresponds to the collection of all joints located on the path between joint *j*  and the root."
Is it correct ? If so, I will fix the doc in this PR.

File: include/pinocchio/algorithm/model.hxx (hunk 2 above)
Comment (florent-lamiraux) [reply include/pinocchio/algorithm/model.hxx]: done

File: include/pinocchio/algorithm/model.hxx (hunk 3 above)
Comment (florent-lamiraux) [reply include/pinocchio/algorithm/model.hxx]: done

**Comment by florent-lamiraux** [comment]
I addressed the above comments, modified test-cpp-model to test the new behavior, and improved the documentation
of Model.supports.

File: include/pinocchio/algorithm/model.hxx [hunk 4]
```diff
@@ -256,12 +273,32 @@ namespace pinocchio
...
+    for (JointIndex jid = frame.parent+1; jid < modelA.joints.size(); ++jid)
+    {
+      if (hasAncestor(modelA, jid, frame.parent))
+      {
+        AJointsBeforeB.push_back(jid);
+      } else
+      {
+        AJointsAfterB.push_back(jid);
+      }
+    }
```
Comment (jmirabel) [inline include/pinocchio/algorithm/model.hxx:295]: Given the compact organization of the list of joints, I would expect that you can copy all the joints of model A until the first joint whose id `id_insert` matches:
- `id_frame_parent < id_insert`: the joint is after the `frame.parent` where modelB is appended.
- `modelA.parents[id_insert] < id_frame_parent`: the joint parent is before `frame.parent`.

`id_insert` is to be added after model B.

If that is true, the code would *arguably* be more atomic.

Comment (jmirabel) [reply include/pinocchio/algorithm/model.hxx:295]: Note that I'm not specifically asking you to change. I'm done with the current implementation.

**Review by jcarpent — APPROVED** [review]
I will merge this PR as it seems fine, even if it can be improved. Let's postpone it as a future work.

File: include/pinocchio/algorithm/model.hxx [hunk 5]
```diff
@@ -256,12 +273,32 @@ namespace pinocchio
...
     details::appendUniverseToModel (modelA, geomModelA, 0, id, model, geomModel);
+    // Compute joints of A that should be added before and after joints of B
+    std::vector<JointIndex> AJointsBeforeB;
+    std::vector<JointIndex> AJointsAfterB;
+    // All joints until the parent of frameInModelA come first
     for (JointIndex jid = 1; jid <= frame.parent; ++jid)
+    {
+      AJointsBeforeB.push_back(jid);
+    }
+    // descendants of the parent of frameInModelA come also before model B
```
Comment (jcarpent) [inline include/pinocchio/algorithm/model.hxx:285]:
```suggestion
    // descendants of the parent of frameInModelA come also before model B
    // TODO(jcarpent): enhancement by taking into account the compactness of the joint ordering.
```

**Pull request merged**